channeldb/migration32/mission_control_store.go

```diff
@@ -99,67 +92,425 @@ func deserializeOldResult(k, v []byte) (*paymentResultOld, error) {
+			uint64(timeReply.UnixNano()),
+		),
+		route: tlv.NewRecordT[tlv.TlvType3, mcRoute](*rt),
+	}
+
+	if success {
+		result.success = tlv.SomeRecordT(
+			tlv.NewRecordT[tlv.TlvType4, lnwire.TrueBoolean](
+				lnwire.TrueBoolean{},
+			),
+		)
 	}
+
+	if failureSourceIdx != nil {
+		result.failureSourceIdx = tlv.SomeRecordT(
+			tlv.NewPrimitiveRecord[tlv.TlvType5, uint8](
+				uint8(*failureSourceIdx),
+			),
+		)
+	}
+
+	if failure != nil {
+		result.failure = tlv.SomeRecordT(
+			tlv.NewRecordT[tlv.TlvType6, failureMessage](
+				failureMessage{failure},
+			),
+		)
+	}
+
+	return result
 }
 
 // paymentResultNew is the information that becomes available when a payment
 // attempt completes.
 type paymentResultNew struct {
-	paymentResultCommon
-	route *mcRoute
+	id               uint64
+	timeFwd          tlv.RecordT[tlv.TlvType1, uint64]
+	timeReply        tlv.RecordT[tlv.TlvType2, uint64]
+	route            tlv.RecordT[tlv.TlvType3, mcRoute]
+	success          tlv.OptionalRecordT[tlv.TlvType4, lnwire.TrueBoolean]
+	failureSourceIdx tlv.OptionalRecordT[tlv.TlvType5, uint8]
+	failure          tlv.OptionalRecordT[tlv.TlvType6, failureMessage]
+}
+
+type failureMessage struct {
+	lnwire.FailureMessage
+}
+
+// Record returns a TLV record that can be used to encode/decode a list of
+// mcRoute to/from a TLV stream.
+func (r *failureMessage) Record() tlv.Record {
+	recordSize := func() uint64 {
+		var (
+			b   bytes.Buffer
+			buf [8]byte
+		)
+		if err := encodeFailureMessage(&b, r, &buf); err != nil {
+			panic(err)
+		}
+
+		return uint64(len(b.Bytes()))
+	}
+
+	return tlv.MakeDynamicRecord(
+		0, r, recordSize, encodeFailureMessage, decodeFailureMessage,
+	)
+}
+
+func encodeFailureMessage(w io.Writer, val interface{}, _ *[8]byte) error {
+	if v, ok := val.(*failureMessage); ok {
+		var b bytes.Buffer
+		err := lnwire.EncodeFailureMessage(&b, v.FailureMessage, 0)
+		if err != nil {
+			return err
+		}
+
+		_, err = w.Write(b.Bytes())
+
+		return err
+	}
+
+	return tlv.NewTypeForEncodingErr(val, "routing.failureMessage")
 }
 
+func decodeFailureMessage(r io.Reader, val interface{}, _ *[8]byte,
+	l uint64) error {
+
+	if v, ok := val.(*failureMessage); ok {
+		msg, err := lnwire.DecodeFailureMessage(r, 0)
+		if err != nil {
+			return err
+		}
+
+		*v = failureMessage{
+			FailureMessage: msg,
+		}
+
+		return nil
+	}
+
+	return tlv.NewTypeForDecodingErr(val, "routing.failureMessage", l, l)
+}
+
+// extractMCRoute extracts the fields required by MC from the Route struct to
+// create the more minimal mcRoute struct.
 // extractMCRoute extracts the fields required by MC from the Route struct to
 // create the more minimal mcRoute struct.
-func extractMCRoute(route *Route) *mcRoute {
+func extractMCRoute(r *Route) *mcRoute {
+	//nolint:lll
 	return &mcRoute{
-		sourcePubKey: route.SourcePubKey,
-		totalAmount:  route.TotalAmount,
-		hops:         extractMCHops(route.Hops),
+		sourcePubKey: tlv.NewRecordT[tlv.TlvType0, Vertex](r.SourcePubKey),
```

Comment:
nit: since we're passing a typed argument into this function, the second type parameter can be inferred by the compiler. So you can remote it to make things a bit more compact (here and in several other places).

Reply:
indeeeed - thanks!